aip/0162.md

```diff
@@ -0,0 +1,229 @@
+---
+aip:
+  id: 162
+  state: draft
+  created: 2019-09-17
+permalink: /162
+redirect_from:
+  - /0162
+---
+
+# Resource Revisions
+
+Some APIs need to have resources with a revision history, where users can
+reason about the state of the resource over time. There are several reasons for
+this:
+
+- Users may want to be able to roll back to a previous revision, or diff
+  against a previous revision.
+- An API may create data which is derived in some way from a resource at a
+  given point in time. In these cases, it may be desirable to snapshot the
+  resource for reference later.
+
+## Guidance
+
+APIs **may** store a revision history for a resource if it is useful to users.
+
+APIs implementing resources with a revision history **must** provide a
+`revision_id` field on the resource:
+
+```proto
+message Book {
+  // The name of the book.
+  string name = 1;
+
+  // Other fields…
+
+  // The revision ID of the book.
+  // This is automatically changed whenever the contents of the book
```

Comment:
It is optional for the revision_id to change whenever any content of the resource changes? E.g. We would like to have a new Model version when it's ML insides got updated, but not when user e.g. added a new label for the current version.

Reply:
I saw below "field must document when new revisions are created", so I understand that the rules may be as we will need it, therefore this comment is taken care of.